src/MagnumPlugins/KtxImporter/KtxImporter.cpp

```diff
@@ -0,0 +1,553 @@
+        std::size_t currentLevel = 0;
+        for(UnsignedInt face = 0; face != numFaces; ++face) {
+            Vector3i mipSize{size};
+
+            /* Load all mipmaps for current face */
+            for(const Implementation::KtxLevel& level: levelIndex) {
+                const std::size_t length = f->compressed
+                    ? imageLength(mipSize, f->pixelFormat.compressed)
+                    : imageLength(mipSize, f->pixelFormat.uncompressed);
+
+                /* Each mipmap byte range contains tightly packed images,
+                   ordered as follows:
+                   layers, faces, (slices, rows, columns) */
+                const std::size_t imageOffset = (layer * numFaces + face) * length;
+                if(level.byteLength < imageOffset + length) {
+                    Error{} << "Trade::KtxImporter::openData(): image data too short, expected at least" << imageOffset + length << "bytes but got" << level.byteLength;
+                    return;
+                }
+
+                f->imageData[layer][currentLevel++] = {mipSize, f->in.suffix(level.byteOffset).slice(imageOffset, imageOffset + length)};
+
+                /* Shrink to next power of 2 */
+                mipSize = Math::max(mipSize >> 1, Vector3i{1});
+            }
+        }
+    }
+
+    /** @todo Read KTXorientation and flip image? What is the default? */
+    /** @todo Read KTXswizzle and apply it. Can't do that for block-compressed formats, just ignore in that case? */
+    /** @todo Read KTXanimData and expose frame time between images through
+              doImporterState (same as StbImageImporter). What if we need
+              doImporterState for something else, like other API format
+              enum values (KTXdxgiFormat__ etc.)? */
+
+    _f = std::move(f);
+}
+
+UnsignedInt KtxImporter::doImage1DCount() const { return (_f->dimensions == 1) ? _f->imageData.size() : 0; }
+
+UnsignedInt KtxImporter::doImage1DLevelCount(UnsignedInt id) { return _f->imageData[id].size(); }
+
+Containers::Optional<ImageData1D> KtxImporter::doImage1D(UnsignedInt id, UnsignedInt level) {
+    const File::LevelData& levelData = _f->imageData[id][level];
+
+    /* Copy image data */
+    Containers::Array<char> data{NoInit, levelData.data.size()};
+    Utility::copy(levelData.data, data);
+
+    /* Endian-swap if necessary */
+    endianSwap(data, _f->typeSize);
+
+    /* Compressed image */
+    if(_f->compressed) {
+        /* Using default CompressedPixelStorage, need explicit one once we
+           support all Vulkan formats */
+        return ImageData1D(_f->pixelFormat.compressed, levelData.size.x(), std::move(data));
+    }
+
+    /* Uncompressed */
+    if(_f->needsSwizzle) swizzlePixels(_f->pixelFormat.uncompressed, data,
+        flags() & ImporterFlag::Verbose ? "Trade::KtxImporter::image2D():" : nullptr);
+
+    /* Rows are tightly packed */
+    PixelStorage storage;
+    storage.setAlignment(1);
```

Comment:
What exactly does `alignment` do? My understanding was that it tells you about possible padding between rows. Since KTX stores rows tightly packed, I thought this should be 1. The only padding KTX has is for the start of an entire mip level, all subsequent images inside that mip level (array layers, cubemap faces) aren't explicitly aligned to anything but the block size, if compressed.

Reply:
> possible padding between rows

Yes, but generally I'm striving for keeping it at 4 when the rows actually *are* 4-byte aligned, because that means less GL state changes when uploading the data, possibly a more optimal pixel copy path chosen in the driver etc.

> KTX stores rows tightly packed

Huh, for some reasons I thought KTX goes a *really long way* to ensure everything is 4-byte aligned, metadata, rows, all that. Probably mixed that up with glTF, sorry. Yes, in this case, can you then set the alignment to 1 only if it's really needed?

Reply:
KTX1 had a row alignment of 4, that might be why it was in the back of your head. KTX2 [says](https://github.khronos.org/KTX-Specification/#levelImages):

> Rows of uncompressed texture images must be tightly packed, equivalent to a GL_UNPACK_ALIGNMENT of 1.

The `DdsImporter` does what you suggested but I didn't realize it was for performance reasons and removed the check, makes perfect sense now 👌